app/controllers/users_controller.rb

```diff
@@ -119,4 +108,29 @@ def update_email
       redirect_to edit_user_path
     end
   end
+
+  # GET /user/discord_account_required?course_requiring_discord
+  def discord_account_required
+    authorize(current_user)
+
+    @course_requiring_discord_account =
+      if params[:course_requiring_discord].present? &&
+           !current_user.discord_account_connected?
+        course =
+          current_user.courses.find_by(id: params[:course_requiring_discord])
+
+        session[:course_requiring_discord] = course.id if course.present?
+        course
+      elsif session.key?(:course_requiring_discord)
+        course =
+          current_user.courses.find_by(id: session[:course_requiring_discord])
+
+        session.delete(:course_requiring_discord)
+        course
+      else
+        redirect_to dashboard_path
```

Comment:
@yash-learner Fixed. I opted to put the `return` statement on L132. I prefer not to write `return redirect_to dashboard_path` because I don't actually intend to `return` whatever value is returned by the call to `redirect_to(dashboard_path)` - I just need to stop execution of the current method.

Also, the boolean returned by the _policy_ method is, by convention, concerned only with the _authorization_ aspect of the request. Is a user _allowed_ to access this page? That answer, for a signed-in user, is always _yes_. But the decision on what should be done / shown when a user accesses an authorized page isn't the policy's concern - for example, to redirect away because there's no reason to show the contents of the page just then.

Reply:
@harigopal 

> The UI has the Return to dashboard button to handle the case where the user has returned to the page after linking account, but the course ID in session has been lost for some reason.

I think it is handled to show `begin_course` when the user visits the page after connecting the `Discord` 

I think after connecting the discord, the user is redirected to `user/discord_account_required` this will cause the code to execute the following piece of code 

```ruby
elsif session.key?(:course_requiring_discord)
  course =
    current_user.courses.find_by(id: session[:course_requiring_discord])

  session.delete(:course_requiring_discord) # we are deleting the key here
  course
```

And in the above code we are deleting the key in the line before, the course. So, when the user now visits the `user/discord_account_required` the user is redirected to `/dashboard` as it falls into else case.

I think if we wish to show the `return to dashboard` when a user visits `user/discord_account_required` then the following code will make sure

```ruby
    @course_requiring_discord_account =
      if params[:course_requiring_discord].present? &&
           !current_user.discord_account_connected?
        course =
          current_user.courses.find_by(id: params[:course_requiring_discord])

        session[:course_requiring_discord] = course.id if course.present?
        course
      elsif session.key?(:course_requiring_discord)
        course =
          current_user.courses.find_by(id: session[:course_requiring_discord])

        session.delete(:course_requiring_discord)
        course
      elsif current_user.discord_account_connected?
        nil
      else
        redirect_to dashboard_path
        return
      end
```

Correct me if I am thinking in a wrong way

Reply:
Why not make the following code 

```ruby
else
  redirect_to dashboard_path
  return
 ```

as

```ruby
else
  redirect_to dashboard_path and return
```

Reply:
Thanks, Got it